File: hdar/archives/hdar_3/hdar/capsule/restore.py
"""Capsule restoration — workspace reconstruction from a sealed capsule.

Restores a workspace from a capsule manifest using the content-addressed
store. Verifies manifest signature, receipt chain, and workspace root hash
after restoration.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Optional, Tuple

from .store import ContentStore, WorkspaceManifest
from .identity import LineageEpoch, deserialize_public_key
from .receipt import ReceiptChain
from .seal import CapsuleManifest, CapsuleSealer

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


class CapsuleRestorer:
    """Restores a workspace from a sealed capsule."""

    def __init__(self, store: ContentStore):
        self.store = store

    def load_capsule(self, capsule_path: Path) -> CapsuleManifest:
        """Load a capsule manifest from a JSON file."""
        data = json.loads(capsule_path.read_text())
        return CapsuleManifest.from_dict(data)

    def verify_manifest(self, manifest: CapsuleManifest, owner_public_key: Ed25519PublicKey) -> bool:
        """Verify the manifest Ed25519 signature and hash."""
        expected_hash = manifest.compute_hash()
        if expected_hash != manifest.manifest_hash:
            return False
        try:
            owner_public_key.verify(
                bytes.fromhex(manifest.signature),
                manifest.canonical_bytes(),
            )
            return True
        except Exception:
            return False

    def verify_receipts(self, manifest: CapsuleManifest, owner_public_key: Ed25519PublicKey) -> bool:
        """Verify the embedded receipt chain using Ed25519 public key."""
        if not manifest.receipts:
            return False
        from .receipt import Receipt
        receipts = [Receipt.from_dict(r) for r in manifest.receipts]
        prev_hash = None
        for r in receipts:
            if r.prior_receipt_hash != prev_hash:
                return False
            if not r.verify(owner_public_key):
                return False
            prev_hash = r.receipt_hash
        return True

    def restore_workspace(
        self,
        manifest: CapsuleManifest,
        dest_dir: Path,
    ) -> Tuple[WorkspaceManifest, bool]:
        """Reconstruct the workspace from the capsule.

        Returns (restored_manifest, hash_matches).
        """
        ws_dict = manifest.workspace_manifest
        if ws_dict is None:
            raise ValueError("capsule has no workspace manifest")

        ws_manifest = WorkspaceManifest.from_dict(ws_dict)

        # Restore files from content store
        self.store.restore_workspace(ws_manifest, dest_dir)

        # Verify restored workspace hash
        restored_manifest = self.store.hash_workspace(dest_dir)
        hash_matches = restored_manifest.root_hash == ws_manifest.root_hash

        return restored_manifest, hash_matches

    def restore(
        self,
        capsule_path: Path,
        dest_dir: Path,
        owner_public_key: Optional[Ed25519PublicKey] = None,
    ) -> dict:
        """Full restoration: load, verify, restore workspace.

        Returns a restoration report dict.
        """
        manifest = self.load_capsule(capsule_path)

        # Signature verification (skip if no key provided)
        sig_valid = None
        if owner_public_key is not None:
            sig_valid = self.verify_manifest(manifest, owner_public_key)

        # Receipt chain verification
        receipts_valid = None
        if owner_public_key is not None:
            receipts_valid = self.verify_receipts(manifest, owner_public_key)

        # Workspace restoration
        restored_manifest, hash_matches = self.restore_workspace(manifest, dest_dir)

        return {
            "agent_id": manifest.agent_id,
            "agent_name": manifest.agent_name,
            "epoch": manifest.epoch,
            "objective": manifest.objective,
            "continuation_point": manifest.continuation_point,
            "working_summary": manifest.working_summary,
            "restoration_contract": manifest.restoration_contract,
            "workspace_root_hash": manifest.workspace_manifest["root_hash"],
            "restored_root_hash": restored_manifest.root_hash,
            "workspace_hash_matches": hash_matches,
            "signature_valid": sig_valid,
            "receipts_valid": receipts_valid,
            "file_count": len(restored_manifest.files),
            "total_size": restored_manifest.total_size,
            "capabilities": manifest.capabilities,
            "capability_note": manifest.capability_note,
            "pending_operations": manifest.pending_operations,
        }

File: hdar/cpp/src/restore.cpp
#include "hdar/restore.hpp"

namespace hdar {

CapsuleRestorer::CapsuleRestorer(ContentStore& store) : store_(store) {}

CapsuleManifest CapsuleRestorer::load_capsule(const std::string& path) const {
    std::string content = read_file_to_string(path);
    JsonValue json = parse_json(content);
    return CapsuleManifest::from_json(json);
}

bool CapsuleRestorer::verify_manifest(const CapsuleManifest& manifest,
                                       const PublicKey& owner_pk) const {
    if (manifest.compute_hash() != manifest.manifest_hash)
        return false;
    return owner_pk.verify_hex(manifest.canonical_bytes(), manifest.signature);
}

bool CapsuleRestorer::verify_receipts(const CapsuleManifest& manifest,
                                       const PublicKey& owner_pk) const {
    if (manifest.receipts.type != JsonValue::Type::Array || manifest.receipts.array_val.empty())
        return false;

    std::optional<std::string> prev_hash;
    for (const auto& rj : manifest.receipts.array_val) {
        Receipt r = Receipt::from_json(rj);
        if (r.prior_receipt_hash != prev_hash)
            return false;
        if (!r.verify(owner_pk))
            return false;
        prev_hash = r.receipt_hash;
    }
    return true;
}

std::pair<WorkspaceManifest, bool> CapsuleRestorer::restore_workspace(
    const CapsuleManifest& manifest, const std::string& dest_dir) const {

    if (!manifest.workspace_manifest)
        throw std::runtime_error("capsule has no workspace manifest");

    auto ws_manifest = WorkspaceManifest::from_json(*manifest.workspace_manifest);
    store_.restore_workspace(ws_manifest, dest_dir);

    auto restored = store_.hash_workspace(dest_dir);
    bool hash_matches = (restored.root_hash == ws_manifest.root_hash);

    return {restored, hash_matches};
}

JsonValue CapsuleRestorer::restore(const std::string& capsule_path,
                                    const std::string& dest_dir,
                                    const PublicKey* owner_pk) const {
    auto manifest = load_capsule(capsule_path);

    bool sig_valid = false;
    bool receipts_valid = false;
    if (owner_pk) {
        sig_valid = verify_manifest(manifest, *owner_pk);
        receipts_valid = verify_receipts(manifest, *owner_pk);
    }

    auto [restored_manifest, hash_matches] = restore_workspace(manifest, dest_dir);

    JsonValue result = JsonValue::object();
    result["agent_id"] = JsonValue::string(manifest.agent_id);
    result["agent_name"] = JsonValue::string(manifest.agent_name);
    result["epoch"] = manifest.epoch;
    result["objective"] = JsonValue::string(manifest.objective);
    result["continuation_point"] = JsonValue::string(manifest.continuation_point);
    result["working_summary"] = JsonValue::string(manifest.working_summary);
    result["restoration_contract"] = JsonValue::string(manifest.restoration_contract);

    if (manifest.workspace_manifest) {
        result["workspace_root_hash"] = JsonValue::string(
            manifest.workspace_manifest->get("root_hash").string_val);
    }
    result["restored_root_hash"] = JsonValue::string(restored_manifest.root_hash);
    result["workspace_hash_matches"] = JsonValue::boolean(hash_matches);
    result["signature_valid"] = JsonValue::boolean(sig_valid);
    result["receipts_valid"] = JsonValue::boolean(receipts_valid);
    result["file_count"] = JsonValue::integer(static_cast<int64_t>(restored_manifest.files.size()));
    result["total_size"] = JsonValue::integer(static_cast<int64_t>(restored_manifest.total_size));
    result["capabilities"] = manifest.capabilities;
    result["capability_note"] = JsonValue::string(manifest.capability_note);
    result["pending_operations"] = manifest.pending_operations;

    return result;
}

} // namespace hdar
